cranelift/codegen/src/isa/zkasm/abi.rs

```diff
@@ -427,12 +424,15 @@ impl ABIMachineSpec for ZkAsmMachineDeps {
                     RegClass::Vector => unimplemented!("Vector Clobber Saves"),
                 };
                 insts.push(Self::gen_store_stack(
-                    StackAMode::SPOffset(cur_offset, ty),
+                    StackAMode::SPOffset(-cur_offset, ty),
```

Comment:
Hm, I seem to remember thinking that it would be better for us to stay as close to how the backend addressing works rather than hacking in the conversion to the stack slots at the very last moment (in the instruction lowering.)

The benefit of doing it that way is that invalid states are simply not representable at all, and any byte->slot conversion logic can occur in the proper boundary of abstraction. If there are any parts of the clift codebase that wants to operate in bytes, the conversion and any decisions on how to set up the slots can occur there (which, if nothing else, brings the potential failure modes closer to their source, which in turn makes e.g. panics actually point at a problem spot, and rather than just "hey your 1 billion lines of assembly is wrong, up to you figure out where!")

Reply:
> I don't see a place where we can intercept this code and do the necessary conversion to slots - what would be the right place to do this?

`gen_load_stack`, perhaps? And then add another, inherent, method for generating stack loads with slot addressing that you could use from within `zkasm` backend.

If nothing else such an approach makes the lifetime of `StackAMode::*` with byte-addressing significantly closer to ε.

Though perhaps your way of doing things is also right – implementing my suggested approach then makes some places have byte-addressed offsets (outside of zkasm) and some other places slot-addressed ones…

Or perhaps (as is often the case) some middle ground is the most appropriate (e.g. we could just slap *another* assert into `gen_load_stack`.)

Reply:
> > I don't see a place where we can intercept this code and do the necessary conversion to slots - what would be the right place to do this?
> 
> `gen_load_stack`, perhaps? And then add another, inherent, method for generating stack loads with slot addressing that you could use from within `zkasm` backend.

This indeed looks very promising, thank you!

I like both of your suggestions about the method with slot addressing and with moving the offset translation to `gen_load_stack` and I think we should implement them.

I would like to take the following route to make sure we're doing this translation correctly:
1. Make sure SHA256 benchmarks works
2. Add more spec tests that exercise this codepath
3. Move to slot addressing as a part of work in https://github.com/near/wasmtime/issues/43

This way I'll have a bit more confidence that this is not what's breaking the SHA256 benchmark and will be able to verify that after this change, the benchmark still works. Does this plan make sense?